feat(core): introduce basic life cycle support for LoopBack applications

raymondfeng opened the pull request:
Implements:
- https://github.com/strongloop/loopback-next/issues/1845
- https://github.com/strongloop/loopback-next/issues/2034

The PR implements the following features:

- Allow registration and actions of life cycle observers using context bindings - See https://github.com/strongloop/loopback-next/blob/lifecycle/docs/site/Life-cycle.md
- Discover and boot life cycle scripts
- Add `lb4 observer` command to generate life cycle scripts - See https://github.com/strongloop/loopback-next/blob/lifecycle/docs/site/Life-cycle-observer-generator.md

## Test `observer` cli

```sh
cd loopback-next
git checkout lifecycle
npm run build
cd sandbox
../packages/cli/bin/cli-main.js app life
cd life
../../packages/cli/bin/cli-main.js observer
cd ../..
npm i
cd sandbox/life
npm start
```

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

davidxiao commented:
when will this PR be merged? :)

raymondfeng commented:
@bajtos Please review.

bajtos commented on packages/core/src/lifecycle.ts:
```diff
@@ -0,0 +1,23 @@
+ */
+export interface LifeCycleAction {
+  start(): Promise<void> | void;
+  stop(): Promise<void> | void;
```
Let's discuss this API. Shouldn't we provide some kind of a context in which the event is happening? Without a context, how can a life-cycle-action handler access relevant data like the application and/or the server being started?

raymondfeng replied:
If we switch to `observer` pattern, yes. For simple `start/stop`, I don't see the need to have an argument for context. Other information can be injected from the application context.

raymondfeng commented on packages/core/src/application.ts:
```diff
@@ -130,6 +131,9 @@ export class Application extends Context {
    */
   public async start(): Promise<void> {
+    await this._forEachComponent(c => {
+      if (isLifeCycleAction(c)) return c.start();
```
Ultimately we'll switch the `observer` pattern so that components can subscribe to the `start/stop` events. See #1972.

Do we want to wait unless we settle on #1972?

bajtos replied:
I think that even with the observer pattern from #1972, we should still allow components to export lifecycle observers in an indirect way, i.e. letting `mountComponent` to handle the actual registration.

It is not clear to me how #1972 would affect the design of lifecycle events and whether the solution based purely on low-level observer APIs would be enough easy to use.

My preferred approach is to use composition instead of inheritance, allow components to export an array of life-cycle listener/observe classes and get this pull request landed.

bajtos replied:
Later on, when #1972 is implemented, we can look into ways how to leverage the observer pattern as a low-level implementation detail to power "LifeCycleObserver".

bajtos commented on packages/core/src/application.ts:
```diff
@@ -130,6 +131,9 @@ export class Application extends Context {
    * @memberof Application
    */
   public async start(): Promise<void> {
+    await this._forEachComponent(c => {
```
This opens an interesting question: should we invoke "start" observers before the servers are started or after?

 - When the observers are invoked before, they have a chance to modify server configuration.
 - When the observers are invoked after, they are able to access the information created during startup, e.g. the port number assigned by the operation system to build the URL of a server listening on `port: 0`.

I think the second use case is more important because I don't see how it could be implemented otherwise? The first use case (modifying config before start) seems redundant to me, because application setup can already be customized during `app.boot()`.

bajtos replied:
Please make sure there is a test capturing this nuance!

raymondfeng replied:
It's always tricky to make sure artifacts are started/stopped in the right order. For example, we usually want servers to be started last and stopped first so that requests can only be handled when the system is ready.

A way to describe the dependency is needed to allow life cycle actions to be triggered by order. For example, each action can declare its dependencies or desired phase.

raymondfeng commented:
@bajtos PTAL

bajtos commented on packages/core/src/keys.ts:
```diff
@@ -38,7 +38,29 @@ export namespace CoreBindings {
+  /**
+   * Binding tag for servers
+   */
+  export const SERVER_TAG = 'server';
```
Instead of using `_TAG` suffix, can we create a new namespace instead?

```ts
CoreBindings.Tags.SERVER
```

bajtos replied:
Or perhaps better: `CoreTags.SERVER`.

So far, namespaces like `CoreBindings`, `RestBindings`, etc. were holding binding keys (addresses) only.

bajtos commented on packages/core/src/keys.ts:
```diff
@@ -66,4 +88,9 @@ export namespace CoreBindings {
+  /**
+   * Binding tag for life cycle observers
+   */
+  export const LIFE_CYCLE_OBSERVER_TAG = 'lifeCycleObserver';
```
Should we keep all tags grouped together?

bajtos commented on packages/core/src/application.ts:
```diff
@@ -139,25 +166,50 @@ export class Application extends Context {
+    debug('Stopping the application...');
+    const bindings = this._findLifeCycleObserverBindings();
+    // Stop in the reverse order
+    for (const binding of bindings.reverse()) {
```
Can we move the call of `.reverse()` out of the `for` loop? E.g.

```ts
const bindings = this._findLifeCycleObserverBindings().reverse();
```

b-admike commented on packages/boot/src/booters/lifecyle-script.booter.ts:
```diff
@@ -0,0 +1,70 @@
+ */
+export const LifeCycleScriptDefaults: ArtifactOptions = {
+  dirs: ['scripts'],
+  extensions: ['.script.js'],
```
`script.ts` as well?

raymondfeng replied:
No, we only scan `*.js` files from `dist`.

bajtos commented on packages/core/src/application.ts:
```diff
@@ -130,7 +147,17 @@ export class Application extends Context {
-    await this._forEachServer(s => s.start());
+    debug('Starting the application...');
+    const bindings = this._findLifeCycleObserverBindings();
+    for (const binding of bindings) {
```
Have you considered extracting this loop into a function accepting 1) a list of bindings 2) a string name of the method to invoke?

```ts
notifyLifeCycleObservers(bindings, 'start');
```

bajtos commented on packages/core/src/application.ts:
```diff
@@ -41,9 +55,11 @@ export class Application extends Context {
+    return this.bind(key)
       .toClass(controllerCtor)
-      .tag('controller');
+      .tag(CoreBindings.CONTROLLER_TAG);
```
It would be great if you could extract the part reworking existing binding tags from strings to constants and send a standalone pull request.

bajtos commented on packages/core/src/application.ts:
```diff
@@ -3,17 +3,31 @@
+  LifeCycleObserver,
+} from './lifecycle';
+import {Server} from './server';
+import debugModule = require('debug');
```
In other places, I was already using `debugFactory` instead of `debugModule`. Let's pick one consistent way going forward please.

bajtos commented on packages/context/src/binding.ts:
````diff
@@ -423,6 +423,22 @@ export class Binding<T = BoundValue> {
+   * For example,
+   * ```ts
+   * const serverTemplate = (binding: Binding) =>
+   *   binding.inScope(BindingScope.SINGLETON).tag('server');
````
Should we use the new `SERVER_TAG` constant instead of a hard-coded string?

bajtos commented on packages/boot/src/booters/lifecyle-script.booter.ts:
```diff
@@ -0,0 +1,70 @@
+ * Default ArtifactOptions for DataSourceBooter.
+ */
+export const LifeCycleScriptDefaults: ArtifactOptions = {
+  dirs: ['scripts'],
```
I find the directory name `scripts` (and also the CLI command `lb4 script`) as not descriptive enough. Every JavaScript source code is essentially a script.

Let's find a name that better conveys the intention please. How about `lifecycle` directory and `lb4 lifecycle`? (`lifecycle-script` seems too long to me.)

bajtos commented on packages/core/src/application.ts:
```diff
@@ -14,13 +14,13 @@ import {Component, mountComponent} from './component';
 import {Server} from './server';
-import debugModule = require('debug');
-const debug = debugModule('loopback:core:application');
+const CoreTags = CoreBindings.Tags;
```
There are quite few  places where we are aliasing `CoreBindings.Tags` to `CoreTags`. I see it as a clear sign that tags should be a new top-level export and not a nested namespace inside `CoreBindings`. I.e.

```ts
import {CoreBindings, CoreTags} from './keys';
```

Thoughts?

bajtos commented:
@raymondfeng I am afraid I have few more concerns about the high-level design.

(1)
Observers are sorted by the framework and servers get a special treatment:
- This is not extensible. When an extension comes with another artifact that needs special position in the list of lifecycle observers, then there is no mechanism allowing the extension to do that. We would have to add another exception to the sorting routine used by core.
- The order of observers is quite non-deterministic for app and especially for extension developers. It's easy to make incorrect assumptions about the order in which observers are executed at one point in application development lifecycle, and then be surprised that this order was changed when a new observer was added via a component or even explicitly by the app.

(2)
> It's always tricky to make sure artifacts are started/stopped in the right order. For example, we usually want servers to be started last and stopped first so that requests can only be handled when the system is ready.

Sure, this makes sense in a lot of cases. However, there are valid use cases where we want to observe the fact that the application has started (all servers are up & listening). Besides the example of printing application's URL to the console, there is also a more compelling scenario - registering the started application with a Service Registry and unregistering it *before* shutdown. 

For example, I can imagine a 3rd-party extension providing start/stop observers that will automatically register the application with a specific Service Registry, e.g. `loopback4-extension-consul` or `loopback4-extension-etcd`. This is not possible to implement with the current design.

I have few ideas for possible solutions, can you come up with few of your own too?
- I am proposing to run all observers in parallel. Conceptually, there is no need to start multiple servers one by one, they can be started in parallel.
- Introduce more lifecycle events. For example, `starting` can start the servers, `started` can register the server with Service Registry. Similarly for `stop`. Alternatively, we can have `beforeStart`, `start` and `afterStart`, but that feels a bit too much to me.
- Support server-specific observers. Instead of waiting for the entire app to start, we can have an observer registered for a RestServer only.
- Don't treat servers as lifecycle observers, but start them explicitly. For example, `app.start` can be implemented as three steps:
  1. Notify all observers - call `start` event (in parallel)
  2. Start all servers
  3. Notify all observers - call `started` event (in parallel)

Anyhow. For me, the important requirement is that we need to give observers the option to decide whether they should run before or after the servers are started. The rest are implementation details...

raymondfeng commented:
@bajtos I introduced a `group` concept to organize observer.  The order of groups can be configured so that we notify them by the order of its group. Observers within the same group are triggered in parallel. For a composite event such as `start`, we emit `preStart`, `start` and `postStart`.

Ultimately such infrastructure should be refactored into https://github.com/strongloop/loopback-next/pull/1972 and possibly align with https://github.com/strongloop/loopback-next/pull/1671.

From user experience perspective, we need to:

1. Allow observers/listeners/handlers to be registered/subscribed against their interests (such as topics/events/phases)
2. Allow observers/listeners/handlers to be discovered for a given event
3. Allow observers/listeners/handlers to be executed in order (for example, by group or phase)
4. Allow observers/listeners/handlers to be unregistered/unsubscribed

bajtos commented on packages/core/src/lifecycle-registry.ts:
```diff
@@ -0,0 +1,211 @@
+
+  /**
+   * Sort the life cycle observer bindings so that we can start/stop them
+   * in the right order. By default, we can start other observers before servers
```
I don't understand why is this sorting still required when we already have `preStart`/`start`/`postStart` groups allowing observers to decide whether they want to run before or after the server is started. Could you please explain? What am I missing?

bajtos commented on packages/core/src/lifecycle-registry.ts:
```diff
@@ -0,0 +1,211 @@
+  protected groupsByOrder: string[] = ['server'];
+
+  setGroupsByOrder(groups: string[]) {
+    this.groupsByOrder = groups || ['server'];
```
I have difficulties to comprehend this group ordering concept and how can application & extension users leverage this configuration point. Could you please explain, perhaps in the form of a documentation as part of this pull request?

raymondfeng commented on packages/context/test/unit/binding.unit.ts:
```diff
@@ -173,6 +173,16 @@ describe('Binding', () => {
     });
   });
 
+  describe('apply(templateFunction)', () => {
```
See https://github.com/strongloop/loopback-next/pull/2119

raymondfeng commented on packages/core/src/lifecycle-registry.ts:
```diff
@@ -0,0 +1,211 @@
+        (binding.type === BindingType.CONSTANT ||
+          binding.scope === BindingScope.SINGLETON) &&
+        (binding.tagMap[CoreTags.LIFE_CYCLE_OBSERVER] != null ||
+          binding.tagMap[CoreTags.SERVER]),
```
We can get rid of `SERVER` tag as we tag servers during `app.server`.

raymondfeng commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,140 @@
+## Add custom life cycle observers by convention
+
+Each application can have custom life cycle observers to be dropped into
+`src/observers` folder as classes implementing `LifeCycleObserver`.
```
I would like to use https://github.com/strongloop/loopback-next/pull/2129

bajtos replied:
> could you expand this section of the documentation and show how lifecycle observes should specify the group they want to belong to? AFAICT, the template used by the CLI tool has been already updated to leverage `@bind()` decorator.

☝️

raymondfeng commented:
> I am proposing to simplify the design as follows:

> Keep an ordered list of named groups (start datasource first, server second), this gives us a lot of flexibility & extensibility.

+1. 

> For each group, support only two events: start and stop. Users can implement post/after hooks by defining new named groups if they like to.

I'm fine to only allow `start` and `stop` for now. IMO, named groups solve the ordering problem, but it might force us to create multiple classes for `preStart`, `start`, and `postStart`. For example:

```
pre-server: [MyPreServer]
server: [MyServer]
post-server: [MyPostServer]
```

With the current design, we allow:
```
server: [MyServer] (allows preStart, start, and postStart in MyServer
```

> Eventually, we should refactor the implementation of an ordered list of named groups into @loopback/phase.

+1.

raymondfeng commented:
BTW, I would like to use https://github.com/strongloop/loopback-next/pull/2122 to simplify the discovery of life cycle observers.

raymondfeng commented:
#2122 is probably implementation detail for this PR. We can merge this one and refine it after #2122 is landed.

bajtos commented:
> I'm fine to only allow start and stop for now. IMO, named groups solve the ordering problem, but it might force us to create multiple classes for preStart, start, and postStart. For example:
>
> ```
> pre-server: [MyPreServer]
> server: [MyServer]
> post-server: [MyPostServer]
> ```
> With the current design, we allow:
> ```
> server: [MyServer] (allows preStart, start, and postStart in MyServer)
> ```

IMO, instead of generic pre-server and post-server groups, we should use (and advocate for) more descriptive names. E.g. `pre-server` can become `server-setup` and `post-server` can become `service-registration` (with Consule/Etcd) or `status-broadcast`.

The most important part is an easy and well understood solution allowing both extensions and app developers to introduce new groups and control their order relatively to other (well known?) groups.

raymondfeng commented:
PTAL

bajtos commented:
I did not manage to find time to review this pull request again 😞 I'll take a look when I get back from vacation in January.

bajtos commented:
I would like us to improve our development process so that more people review foundational features like the one proposed in this pull request. I have to goals in mind:
 - Get more opinions on the design & architecture decisions we are making - more eyes see more.
 - Spread the knowledge about how things are designed and implemented, so that people making subsequent changes can get meaningful feedback from more people than just me and @raymondfeng 

In that light, I'd kindly like to ask members of @strongloop/loopback-maintainers, especially (but not only) @hacksparrow and @jannyHou, to review the changes proposed in this pull request. Please don't focus on coding style and implementation details only, try to build understanding of the big picture and the high-level design. If there are anything that's not clear to you from the documentation, pull request discussion and/or the code, then please ask.

Few suggestions:
- Start at high level and ask yourself: do you understand the use cases/scenarios the PR is addressing?
- Put yourself in the shoes of LB4 users building applications. Would you find the proposed feature and APIs easy to discover and use? Do they make it easy to write application code that's clean, brief and easy to understand?
- Do a similar assessment from the point of view of extension developers.
- Think about our existing conventions and design decisions made in the past. Are the newly proposed APIs consistent with the rest of the framework?
- Implementation wise: is the code easy to read, understand and reason about?
- Testing: are all major scenarios and edge cases covered by tests? Are the individual tests easy to understand? Can you tell what is being tested by each individual test? Will the test print a helpful assertion error on failure?

jannyHou commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,212 @@
+
+- datasource
+
+  - connect/disconnect mongodb
```
the tag here is a little bit weird, did you mean something like
```
- connect/disconnect
- mongodb
- mysql
```
?

hacksparrow commented on packages/cli/generators/observer/templates/observer-template.ts.ejs:
```diff
@@ -0,0 +1,33 @@
+import {bind} from '@loopback/context';
+import {
+  /* inject, Application, */
```
Let's add `CoreBindings`, since the commented code below (`CoreBindings.APPLICATION_INSTANCE`) refers to it.

bajtos commented on packages/core/src/application.ts:
```diff
@@ -140,18 +148,37 @@ export class Application extends Context {
+   * @param keyNamespace Binding key prefix
+   * @param tag Tag name
+   */
+  private _findByNamespaceOrTag(keyNamespace: string, tag?: string) {
```
Great idea, +1 from me

hacksparrow commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,212 @@
+
+- Components
+
+  - A component can register life cycle observers
```
"life cycle observers" or "life cycle events"? If "life cycle observers", can only components register them?

hacksparrow commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,212 @@
+ * Observers to handle life cycle start/stop events
+ */
+export interface LifeCycleObserver {
+  preStart?(): ValueOrPromise<void>;
```
Wouldn't a dynamic interface (no hard-coded methods) be more flexible? Or do we plan support events only around start and stop?

Some other events that I can think of "ready", "connected", "online" (which was "offline" a while ago), "message"; especially if a component is dependent on other components asynchronously.

Something to think about, since we are considering LB4 as a messaging hub too (can't find your PR).

This makes we wonder if a "global event listener" is what we are looking for.

The interface of `LifeCycleObserver` is "set in stone", but with a "global event listener", we have the flexibility to accommodate an infinite type of events, not just those based around start and stop.

hacksparrow commented:
The concept seems to be based around the start and stop events of the main Application. How about we make it a general purpose implementation? That way we will be able to listen for lifecycle events on other components as well.

@raymondfeng

hacksparrow commented:
Making it a general purpose implementation will make "[Adding logic to models](https://loopback.io/doc/en/lb3/Adding-logic-to-models.html)" (maybe LB4 has a different approach, just an example), and other event/situation based handlers a lot easier.

bajtos commented:
> The concept seems to be based around the start and stop events of the main Application. How about we make it a general purpose implementation? That way we will be able to listen for lifecycle events on other components as well.

@hacksparrow see https://github.com/strongloop/loopback-next/pull/1928#discussion_r230709564

This pull request is implementing start/stop events only to avoid over-engineering and have a specific use case to focus on. I believe the pull request https://github.com/strongloop/loopback-next/pull/1972 is implementing a more generic infrastructure that can be used for other purposes, the plan is to rework start/stop implementation on top of that PR once it lands.

hacksparrow commented:
Thanks for the details, @bajtos. As long as we eventually switch this implementation to be based on #1972, I am fine with it. Will be looking deeper into the implementation details today.

hacksparrow commented:
@bajtos @raymondfeng we should have a unified means of event notification. The implementation in this PR and #2122 are different, although the basic requirements are the same; this is a hint at fragmentation and impending technical debt.

I am inclined towards making https://github.com/strongloop/loopback-next/pull/1972 a priority and landing it first.

This PR and #2122 should be using a common low-level notification infrastructure.

hacksparrow commented:
@bajtos is there a reason why we might want to land this implementation as-is, without waiting for #1972?

hacksparrow commented:
> The worst outcome of our discussion would be to combine this pull request and #1972 into one huge change-set.

Nah, #1972 should land on its own. This PR should then be changed to use the infrastructure created by #1972, so should #2122 be. The size of the new PRs would then be smaller than the current ones, since the common implementation details would reside somewhere else.

@raymondfeng thoughts? I can actively help to land #1972.

hacksparrow commented:
Another place where we would be using #1972 - https://github.com/strongloop/loopback-next/issues/1919.

raymondfeng commented:
@hacksparrow I wanted to land in the following order:
- #2122: make it simpler to implement extension point/extension pattern.
- #2259: make it simpler to configure bound services
- #2249: illustrate the usage of #2122 and #2259 
- #1928: organize life cycle observers as extensions
- #1972: extract the common code from #2122, #1928 as refinement

hacksparrow commented:
Thanks @raymondfeng for the update.

bajtos commented:
@raymondfeng what's the status of this pull request? Is it ready for another round of reviews?

Going forward, could you please use GitHub's recently added feature [Re-request review on a pull request](https://github.blog/changelog/2019-02-21-re-request-review-on-a-pull-request/) to let reviewers know that it's time for another review?

bajtos commented:
@raymondfeng AFAICT from what GitHub is showing to me, the following comments were not addressed yet: 

https://github.com/strongloop/loopback-next/pull/1928#issuecomment-451933919
(1)
@raymondfeng in https://github.com/strongloop/loopback-next/pull/1928#issuecomment-445047426, you wrote:

> I'm fine to only allow start and stop for now.

However, the pull request still uses preStart/start/postStart hooks. Is it intentional? What's the reasoning behind the decision?

(2)
Also AFAICT, the discussion in https://github.com/strongloop/loopback-next/pull/1928#discussion_r230709564 has not been resolved yet. Besides the important design decision we need to make, it reminds me of a missing documentation for extension authors.

(3)
Please add a new page nested under https://loopback.io/doc/en/lb4/Extending-LoopBack-4.html and show to extension developers what is the idiomatic way for an extension to be notified about lifecycle events. As I see it, the page `docs/site/Life-cycle.md` is focused on application developers only, which is good and needed, but we also need docs focused on extension developers.

https://github.com/strongloop/loopback-next/pull/1928#pullrequestreview-189799320
(4)
And few more minor comments related to implementation details:

https://github.com/strongloop/loopback-next/pull/1928#discussion_r245646461
Now that #2129 has landed, could you expand this section of the documentation and show how lifecycle observes should specify the group they want to belong to? AFAICT, the template used by the CLI tool has been already updated to leverage @bind() decorator.

https://github.com/strongloop/loopback-next/pull/1928#discussion_r245646804
Should the CLI ask for the observer group via an interactive prompt and then set the group here?

https://github.com/strongloop/loopback-next/pull/1928#discussion_r245647310
I believe we are using this form in other places:

```ts
import * as debugFactory from 'debug';
```

Let's stay consistent please.

https://github.com/strongloop/loopback-next/pull/1928#discussion_r245647743
Please make the patch smaller by preserving the place where start/stop methods are defined.

You pull request is moving them from lines 136-172 to lines 200-218.

https://github.com/strongloop/loopback-next/pull/1928#discussion_r245647995
Consider extracting this line into a shared helper function, e.g.

```ts
    const lifecycle = await this.getLifeCycleObserverRegistry();
```

---

I see also more comments from other team members that does not seem to be addressed or responded to. Please review them all.

bajtos commented:
It's annoying that GitHub is hiding away comments that were not resolved, please keep in mind that you need to expand the hidden comments:

<img width="814" alt="Screen Shot 2019-03-21 at 13 19 20" src="https://user-images.githubusercontent.com/1140553/54751694-0cbb4a80-4bdc-11e9-96a7-575bd7cdd139.png">

raymondfeng commented:
> #1928 (comment)
(1)
@raymondfeng in #1928 (comment), you wrote:
>> I'm fine to only allow start and stop for now.

> However, the pull request still uses preStart/start/postStart hooks. Is it intentional? What's the reasoning behind the decision?

Fixed.

> (2)
Also AFAICT, the discussion in #1928 (comment) has not been resolved yet. Besides the important design decision we need to make, it reminds me of a missing documentation for extension authors.

Added Component.lifeCycleObservers as an array of observer classes.

> (3)
Please add a new page nested under https://loopback.io/doc/en/lb4/Extending-LoopBack-4.html and show to extension developers what is the idiomatic way for an extension to be notified about lifecycle events. As I see it, the page docs/site/Life-cycle.md is focused on application developers only, which is good and needed, but we also need docs focused on extension developers.

TBD.

> #1928 (review)
(4)
And few more minor comments related to implementation details:

> #1928 (comment)
Now that #2129 has landed, could you expand this section of the documentation and show how lifecycle observes should specify the group they want to belong to? AFAICT, the template used by the CLI tool has been already updated to leverage @Bind() decorator.

Fixed.

> #1928 (comment)
Should the CLI ask for the observer group via an interactive prompt and then set the group here?

Fixed.

> #1928 (comment)
I believe we are using this form in other places:
import * as debugFactory from 'debug';
Let's stay consistent please.

Fixed.

> #1928 (comment)
Please make the patch smaller by preserving the place where start/stop methods are defined.

See #2625

> You pull request is moving them from lines 136-172 to lines 200-218.

Fixed.

> #1928 (comment)
Consider extracting this line into a shared helper function, e.g.
    const lifecycle = await this.getLifeCycleObserverRegistry();

Fixed.

emonddr commented on packages/core/src/__tests__/unit/application-lifecycle.unit.ts:
```diff
@@ -0,0 +1,149 @@
+
+      // The app.start should not attempt to start this binding.
+      app.bind('controllers.servers').to({});
+      await app.start();
```
I had the same request. The test is passing with app.start() and then app.stop() and no assertions.

bajtos replied:
+1

raymondfeng commented on docs/site/Life-cycle.md:
````diff
@@ -0,0 +1,282 @@
+app.lifeCycleObserver(MyObserver);
+```
+
+Please note that `app.server()` automatically registers servers as life cycle
````
Yes.

raymondfeng commented on docs/site/Life-cycle.md:
````diff
@@ -0,0 +1,282 @@
+For example,
+
+```ts
+app
````
Use `@bind` decoration.

raymondfeng commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,282 @@
+For example, with a group order as `['datasource', 'server']` and three
+observers registered as follows:
+
+- datasource group: MySQLDataSource, MongoDBDataSource
```
There is no ordering within a group. It's probably by the order of registration.

raymondfeng commented on packages/core/src/__tests__/unit/lifecycle-registry.unit.ts:
```diff
@@ -0,0 +1,99 @@
+      .bind(CoreBindings.LIFE_CYCLE_OBSERVER_REGISTRY)
+      .toClass(LifeCycleObserverRegistry)
+      .inScope(BindingScope.SINGLETON);
+    registry = await context.get(CoreBindings.LIFE_CYCLE_OBSERVER_REGISTRY);
```
No. Please note `CoreBindings.LIFE_CYCLE_OBSERVER_REGISTRY` is already strongly-typed.

raymondfeng commented on packages/core/src/__tests__/unit/lifecycle-registry.unit.ts:
```diff
@@ -0,0 +1,99 @@
+    registry = await context.get(CoreBindings.LIFE_CYCLE_OBSERVER_REGISTRY);
+  }
+
+  function givenObserver(name: string, group = '') {
```
Yes.

raymondfeng commented on packages/core/src/lifecycle-registry.ts:
```diff
@@ -0,0 +1,241 @@
+  /**
+   * Notify observers of the same group in parallel, default to `true`
+   */
+  parallel?: boolean;
```
I don't think `type` allows initializer.

raymondfeng commented on packages/core/src/lifecycle-registry.ts:
```diff
@@ -0,0 +1,241 @@
+  protected getObserverGroupsByOrder(): LifeCycleObserverGroup[] {
+    const bindings = this.observersView.bindings;
+    const groups = this.sortObserverBindingsByGroup(bindings);
+    if (debug.enabled) {
```
It's controlled by `DEBUG` env var. Checking `debug.enabled` to avoid unnecessary calculation if the debug statement uses parameters with expressions.

raymondfeng commented on docs/site/Life-cycle.md:
````diff
@@ -0,0 +1,282 @@
+2. lb4 observer
+
+```
+? Observer name: Hello
````
We added.

raymondfeng commented:
> Please add a new page nested under https://loopback.io/doc/en/lb4/Extending-LoopBack-4.html and show to extension developers what is the idiomatic way for an extension to be notified about lifecycle events. As I see it, the page docs/site/Life-cycle.md is focused on application developers only, which is good and needed, but we also need docs focused on extension developers.

Added.

bajtos commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,217 @@
+
+We can then configure the application to trigger observers group by group as
+configured by an array of groups in order such as `['datasource', 'server']`.
+Observers within the same group can be notified in parallel.
```
> Can be notified in parallel

I find this part ambiguous. Are you trying to say that observers are notified in series by default, but there is an option to enable parallel processing instead? How can readers of this paragraph find how to set that option?

Or did you perhaps want to say that _observers within the same group **are** notified in parallel_?

bajtos commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,217 @@
+  // ...
+}
+
+app.add(createBindingFromClass(MyObserver));
```
Probably out of scope of this pull request: would it make sense to introduce Context-level sugar API for adding a binding created from a class?

```diff
- app.add(createBindingFromClass(MyObserver));
+ app.bindFromClass(MyObserver);
```

raymondfeng replied:
+1.

bajtos commented on docs/site/Life-cycle.md:
````diff
@@ -0,0 +1,217 @@
+```
+
+Observers are sorted using `groupsByOrder` as the relative order. If an observer
+is tagged with a group that are not in `groupsByOrder`, it will come before any
````
```suggestion
is tagged with a group that is not in `groupsByOrder`, it will come before any
```

bajtos commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,217 @@
+  '1-custom-group': ['my-observer-3'],
+  '2-custom-group': ['my-observer-4'],
+  'g1': ['my-observer-1'],
+  'g2': ['my-observer-2'],
```
This example would work for multiple different sorting rules. For example, `1-custom-group` is alphabetically preceding `g1`, also `g1` is before `g2`. Your entire list is sorted in the alphabetical order, it's not clear which part is controller by specific rules for unknown groups, which part is controlled by `groupsByOrder` and which is controlled by alphabetical sorting.


I am proposing to use the following example instead:

- `groupsByOrder `is set to `['setup-servers', 'announce-services']`
- 'my-observer-1' in group 'setup-servers'
- 'my-observer-2' in group 'announce-services'
- 'my-observer-4' in 'my-custom-group'
- 'my-observer-3' in 'another-custom-group'

Also please check how is this section rendered from Markdown to HTML, I think it may need a bit of polishment.

bajtos commented on docs/site/Life-cycle.md:
````diff
@@ -0,0 +1,217 @@
+```
+
+The execution order of observers within the same group is not defined. If you
+want to have one to be invoked before the other, mark them with two groups.
````
👍

bajtos commented on packages/core/src/__tests__/unit/application-lifecycle.unit.ts:
```diff
@@ -0,0 +1,148 @@
+  });
+});
+
+class FakeComponent implements Component {
```
What aspects of this component make it a fake component? To me, this looks like a legitimate component implementing LifeCycleObserver interface and providing two "dummy" servers.

```suggestion
class ObservingComponentWithServers implements Component, LifeCycleObservers {
```

bajtos replied:
> What aspects of this component make it a fake component? To me, this looks like a legitimate component implementing LifeCycleObserver interface and providing two "dummy" servers.

@raymondfeng PTAL ☝️

bajtos commented on packages/core/src/__tests__/unit/application-lifecycle.unit.ts:
```diff
@@ -0,0 +1,148 @@
+  }
+}
+
+class FakeServer extends Context implements Server {
```
Similarly here.

We are using Sinon to build test doubles in other places. If you are using "fake" as a synonym to a test double, then please use the name `ServerStub` instead, it's more consistent with Sinon terminology.

Personally, I'd use a name that conveys the intention to observe lifecycle events, e.g. `ObservingServer`.

bajtos replied:
☝️ PTAL

bajtos commented on packages/core/src/__tests__/unit/application-lifecycle.unit.ts:
```diff
@@ -0,0 +1,148 @@
+
+    it('does not attempt to start poorly named bindings', async () => {
+      const app = new Application();
+      app.component(FakeComponent);
```
I don't see how is FakeComponent relevant to this particular test case. Can you remove this line please?

bajtos commented on packages/core/src/application.ts:
```diff
@@ -216,6 +222,26 @@ export class Application extends Context {
+      namespace: CoreBindings.LIFE_CYCLE_OBSERVERS,
+      type: CoreTags.LIFE_CYCLE_OBSERVER,
+      defaultScope: BindingScope.SINGLETON,
+    }).apply(asLifeCycleObserver);
```
Can we move the default namespace, type and defaultScope to `asLifeCycleObserver` template?

```ts
const binding = createBindingFromClass(ctor, {name})
  .apply(asLifeCycleObserver);
```

bajtos replied:
Or even better:

```ts
app.bindFromClass(ctor, {name}, asLifeCycleObserver);
```

raymondfeng replied:
> Can we move the default namespace, type and defaultScope to asLifeCycleObserver template?

In fact, no. Such information is used to infer binding attributes but the template function is applied after the binding is created.

bajtos replied:
Does it mean that if the lifecycle class is decorated with `@bind` and sets the type to something else than `CoreTags.LIFE_CYCLE_OBSERVER`, then the result of calling `app.lifeCycleObserver` is a binding with a type that's not LIFE_CYCLE_OBSERVER? Similarly for `namespace`. Are we okay with that? I find it surprising to be honest.

bajtos commented on packages/core/src/lifecycle-registry.ts:
```diff
@@ -0,0 +1,243 @@
+    for (const group of groups) {
+      const observersForGroup: LifeCycleObserver[] = [];
+      const bindingsInGroup = reverse
+        ? group.bindings.reverse()
```
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reverse
> The `reverse()` method reverses an array in place. The first array element becomes the last, and the last array element becomes the first.

I don't think it's a good idea to modify `group.bindings` array.

```suggestion
        ? [...group.bindings].reverse()
```

Please add a test that fails in the current implementation and passes with the fix in place.

bajtos commented on packages/core/src/lifecycle.ts:
```diff
@@ -0,0 +1,63 @@
+ * Find all life cycle observer bindings. By default, a binding tagged with
+ * `CoreTags.LIFE_CYCLE_OBSERVER`
+ */
+export const lifeCycleObserverFilter: BindingFilter = binding =>
```
Let's use a regular named function please.

```ts
export function lifeCycleObserverFilter(binding: Binding): boolean {
  // ...
}
```

raymondfeng replied:
I was trying to make it clear that `lifeCycleObserverFilter` is a `BindingFilter` function.

hacksparrow commented:
@bajtos @raymondfeng  Is there a reason why `packages/boot/test` hasn't been moved to `packages/boot/src/__test__`?

hacksparrow commented on docs/site/Extension-life-cycle.md:
````diff
@@ -0,0 +1,79 @@
+
+```ts
+export class MyComponentWithObservers {
+  lifeCycleObservers = [XObserver, YObserver];
````
@bajtos your thoughts?

raymondfeng replied:
I'm adding `implements Component` and some comments to clarify.

raymondfeng commented on docs/site/Life-cycle.md:
````diff
@@ -0,0 +1,237 @@
+
+```ts
+const registry = await app.get(CoreBindings.LIFE_CYCLE_OBSERVER_REGISTRY);
+registry.setGroupsByOrder(['g1', 'g2', 'server']);
````
Good idea. Changed.

raymondfeng commented:
> @bajtos @raymondfeng Is there a reason why packages/boot/test hasn't been moved to packages/boot/src/__test__?

Good catch. The PR was created before we moved test to `src/__tests__` and these are left-over files. Fixed now.

raymondfeng commented:
@bajtos PTAL.

bajtos commented:
@raymondfeng I don't have bandwidth to review this patch today, but I do want to check your updates before this patch is merged. Please give me few more days.

bajtos commented on packages/core/src/__tests__/unit/application-lifecycle.unit.ts:
```diff
@@ -13,6 +19,7 @@ import {
   LifeCycleObserver,
   Server,
 } from '../..';
+import {lifeCycleObserver} from '../../lifecycle';
```
Please import from `../..`, just add the new imported symbol to the existing import above.

bajtos commented on packages/core/src/__tests__/unit/application-lifecycle.unit.ts:
```diff
@@ -153,6 +198,37 @@ class MyObserver implements LifeCycleObserver {
+  },
+  scope: BindingScope.SINGLETON,
+})
+class MyObserverWithBind implements LifeCycleObserver {
```
This class is used from a single test only, please move its definition to that test case. I think it will make the test easier to read.

bajtos commented on packages/core/src/__tests__/unit/application-lifecycle.unit.ts:
```diff
@@ -153,6 +198,37 @@ class MyObserver implements LifeCycleObserver {
+}
+
+@lifeCycleObserver('my-group', BindingScope.SINGLETON)
+class MyObserverWithDecorator implements LifeCycleObserver {
```
This class is used from a single test only, please move its definition to that test case. I think it will make the test easier to read.

raymondfeng replied:
No, it's used in more than one tests.

bajtos commented on packages/core/src/lifecycle.ts:
```diff
@@ -57,3 +65,17 @@ export function asLifeCycleObserver<T = unknown>(binding: Binding<T>) {
+ * @param group Optional observer group name
+ * @param scope Optional binding scope
+ */
+export function lifeCycleObserver(group = '', scope?: BindingScope) {
```
I am concerned about flexibility of this API. Have you considered to accept the same arguments as `@bind` does?

```ts
export function lifeCycleObserver(group = '', ...specs: BindingSpec[]) {
  return bind(
    asLifeCycleObserver,
    {tags: [CoreTags.LIFE_CYCLE_OBSERVER_GROUP]: group},
    ...specs
  );
}
```

raymondfeng replied:
👍

bajtos commented on docs/site/Extension-life-cycle.md:
```diff
@@ -0,0 +1,79 @@
+      [CoreTags.LIFE_CYCLE_OBSERVER_GROUP]: 'g1',
+    },
+  },
+  asLifeCycleObserver,
```
I find this example pretty complex. Can we show the new `lifeCycleObserver` decorator instead?

bajtos commented on docs/site/Life-cycle-observer-generator.md:
````diff
@@ -0,0 +1,83 @@
+Adds a new [LifeCycleObserver](Life-cycle.md) class to a LoopBack application.
+
+```sh
+lb4 observer [--group <group>] [<name>]
````
Please describe the new generator in CLI README too, see https://github.com/strongloop/loopback-next/blob/master/packages/cli/README.md

raymondfeng replied:
Good catch.

raymondfeng commented:
> Does it mean that if the lifecycle class is decorated with @bind and sets the type to something else than CoreTags.LIFE_CYCLE_OBSERVER, then the result of calling app.lifeCycleObserver is a binding with a type that's not LIFE_CYCLE_OBSERVER? Similarly for namespace. Are we okay with that? I find it surprising to be honest.

Please note that `type` and `namespace` options in `createBindingFromClass()` overrides `@bind`.

hacksparrow commented on docs/site/Extension-life-cycle.md:
````diff
@@ -0,0 +1,71 @@
+Life cycle observers can be registered via a component too:
+
+```ts
+export class MyComponentWithObservers {
````
@raymondfeng ` implements Component` not added.

hacksparrow replied:
The following code doesn't make sense in the context of the statement "_Life cycle observers can be registered via a component too:_" There's no registration going on.

hacksparrow commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,256 @@
+   * Populate `lifeCycleObservers` per `Component` interface to register life
+   * cycle observers
+   */
+  lifeCycleObservers = [XObserver, YObserver];
```
The current documentation does not contain any information about `lifeCycleObservers`. Opened an issue to document Components - https://github.com/strongloop/loopback-next/issues/2701

hacksparrow commented on packages/boot/src/__tests__/fixtures/lifecycle-observer.artifact.ts:
```diff
@@ -0,0 +1,18 @@
+    this.status = 'started';
+  }
+
+  stop() {
```
Any reason for not making this async as well?

raymondfeng replied:
Added some comments in the code.

bajtos commented:
> > Does it mean that if the lifecycle class is decorated with @bind and sets the type to something else than CoreTags.LIFE_CYCLE_OBSERVER, then the result of calling app.lifeCycleObserver is a binding with a type that's not LIFE_CYCLE_OBSERVER? Similarly for namespace. Are we okay with that? I find it surprising to be honest.

> Please note that `type` and `namespace` options in `createBindingFromClass()` overrides `@bind`.

Well in that case, what are the arguments against setting the tag & namespace via the `asLifeCycleObserver` template?

What I am looking for: an easy way how to bind a life-cycle observer without using `createBindingFromClass` or `app.lifeCycleObserver`.

bajtos commented on packages/core/src/application.ts:
```diff
@@ -77,12 +90,13 @@ export class Application extends Context {
       type: CoreTags.SERVER,
       defaultScope: BindingScope.SINGLETON,
-    });
+    }).apply(asLifeCycleObserver);
```
A typical LifeCycleObserver is bound with the following metadata

```ts
    const binding = createBindingFromClass(ctor, {
	      name,
	      namespace: CoreBindings.LIFE_CYCLE_OBSERVERS,
	      type: CoreTags.LIFE_CYCLE_OBSERVER,
	      defaultScope: BindingScope.SINGLETON,
	    }).tag(CoreTags.LIFE_CYCLE_OBSERVER);
```

When a server is registered as an observer, some of the metadata will be different:

- namespace: `SERVERS` vs. `LIFE_CYCLE_OBSERVERS`
- type: `SERVER` vs. `LIFE_CYCLE_OBSERVER`

Is it intended and acceptable? I am concerned about possible confusion it may create.

raymondfeng replied:
The only requirement is to be tagged as `CoreTags.LIFE_CYCLE_OBSERVER`. I have updated code accordingly.

bajtos commented on packages/core/src/application.ts:
```diff
@@ -216,6 +222,26 @@ export class Application extends Context {
+      namespace: CoreBindings.LIFE_CYCLE_OBSERVERS,
+      type: CoreTags.LIFE_CYCLE_OBSERVER,
+      defaultScope: BindingScope.SINGLETON,
+    }).tag(CoreTags.LIFE_CYCLE_OBSERVER);
```
Can we use `.apply(asLifeCycleObserver)` here? Would the following work?

```suggestion
    }).apply(asLifeCycleObserver)
```

hacksparrow commented on docs/site/Extension-life-cycle.md:
```diff
@@ -0,0 +1,71 @@
+### Implement a life cycle observer
+
+A life cycle observer class optionally implements `start` and `stop` methods to
+be invoked upon `start` and `stop` events.
```
It'll add to the clarity if it is mentioned `start` and `stop` events of what.

hacksparrow commented on docs/site/Life-cycle.md:
```diff
@@ -0,0 +1,256 @@
+
+- DataSources
+
+  - connect: Connect to the underlying database or service
```
Are these events supported? If yes, we need to show an example.

hacksparrow commented on docs/site/Extension-life-cycle.md:
```diff
@@ -0,0 +1,71 @@
+
+As described in [Life cycle](Life-cycle.md), a LoopBack
+[Application](Application.md) has its own life cycles at runtime. Corresponding
+events such as `start` and `stop` are emitted upon the state change.
```
Are there any events apart from `start` and `stop` that can be observed? If no, should make it very clear about the fact.

raymondfeng commented:
The observer pattern can be used to implement life cycle observers but the proposed apis can stay. If we have to change, we can always release a new major version. Release often, release fast :-)
